src/FluffySpoon.AspNet.LetsEncrypt/LetsEncryptChallengeApprovalMiddleware.cs

```diff
@@ -1,56 +1,57 @@
-﻿using System;
-using System.Linq;
+﻿using System.Linq;
 using System.Threading.Tasks;
 using FluffySpoon.AspNet.LetsEncrypt.Persistence;
 using Microsoft.AspNetCore.Http;
 using Microsoft.Extensions.Logging;
 
 namespace FluffySpoon.AspNet.LetsEncrypt
 {
-	public class LetsEncryptChallengeApprovalMiddleware : ILetsEncryptChallengeApprovalMiddleware
-	{
-		private readonly RequestDelegate _next;
-		private readonly ILogger<ILetsEncryptChallengeApprovalMiddleware> _logger;
-
-		private readonly IPersistenceService _persistenceService;
-
-		public LetsEncryptChallengeApprovalMiddleware(
-			RequestDelegate next,
-			ILogger<ILetsEncryptChallengeApprovalMiddleware> logger,
-			IPersistenceService persistenceService)
-		{
-			_next = next;
-			_logger = logger;
-			_persistenceService = persistenceService;
-		}
-
-		public async Task InvokeAsync(
-			HttpContext context)
-		{
-			var path = context.Request.Path.ToString();
-
-			const string magicPrefix = "/.well-known/acme-challenge/";
-			if (path.StartsWith(magicPrefix))
-			{
-				_logger.LogDebug("Challenge invoked: {challengePath}", path);
-
-				var requestedToken = path.Substring(magicPrefix.Length);
-
-				var allChallenges = await _persistenceService.GetPersistedChallengesAsync();
-				var matchingChallenge = allChallenges.FirstOrDefault(x => x.Token == requestedToken);
-				if (matchingChallenge == null)
-				{
-					_logger.LogInformation("The given challenge did not match {challengePath} among {allChallenges}", path, allChallenges);
-
-					await _next(context);
-					return;
-				}
-
-				await context.Response.WriteAsync(matchingChallenge.Response);
-				return;
-			}
-
-			await _next(context);
-		}
-	}
+    public class LetsEncryptChallengeApprovalMiddleware : ILetsEncryptChallengeApprovalMiddleware
+    {
+        private const string MagicPrefix = "/.well-known/acme-challenge/";
+
+        private readonly RequestDelegate _next;
+        private readonly ILogger<ILetsEncryptChallengeApprovalMiddleware> _logger;
+        private readonly IPersistenceService _persistenceService;
+
+        public LetsEncryptChallengeApprovalMiddleware(
+            RequestDelegate next,
+            ILogger<ILetsEncryptChallengeApprovalMiddleware> logger,
+            IPersistenceService persistenceService)
+        {
+            _next = next;
+            _logger = logger;
+            _persistenceService = persistenceService;
+        }
+
+        public Task InvokeAsync(HttpContext context)
+        {
+            var path = context.Request.Path;
+            if (path.HasValue && path.Value.StartsWith(MagicPrefix))
```

Comment:
I think you could reduce the allocations even further if you got rid of the `path` variable and instead used this statement:

`if(context.Request.Path?.StartsWith(MagicPrefix) == true)`

Reply:
`Path` here is an instance of PathString value type it does not provide nullable semantics so `?.` is not an option, unfortunately.

I believe compiler + runtime would eliminate the `path` local variable out of the native stack space here and put it on some CPU registry anyway.

Reply:
Looking at the sources though it is possible to write `context.Request.Path.Value?.StartsWith(MagicPrefix) == true` but I have a gut feeling that the native code would be worse or same in the matter of perf. We could check via Benchmark.Net if you are curious.